src/Init/Data/BitVec/Bitblast.lean

```diff
@@ -1750,6 +1750,152 @@ theorem toInt_srem (x y : BitVec w) : (x.srem y).toInt = x.toInt.tmod y.toInt :=
         ((not_congr neg_eq_zero_iff).mpr hyz)]
       exact neg_le_intMin_of_msb_eq_true h'
 
+@[simp]
+theorem msb_neg_of_msb_false {x : BitVec w} (hx : x.msb = false) :
+    (-x).msb = decide (x ≠ 0) := by
+  by_cases hw : w = 0; subst hw; decide +revert
+  have wpos : 0 < w := by omega
+  simp only [msb_neg, hx, bne_false]
+  simp only [bool_to_prop]
+  simp [hx, wpos]
+
+@[simp]
+theorem msb_neg_of_msb_true {x : BitVec w} (hx : x.msb = true) :
+    (-x).msb = decide (x = intMin w) := by
+  simp only [msb_neg, hx, bne_true, Bool.not_and]
+  simp only [bool_to_prop]
+  simp [hx]
+
+theorem msb_neg_umod_neg_of_msb_true_of_msb_true
+    {x y : BitVec w} (hx : x.msb = true) (hy : y.msb = true) :
+      (-x % -y).msb = (decide (x = intMin w) && decide (-x < -y)) := by
+  simp only [msb_umod, msb_neg_of_msb_true, hx]
+  simp only [bool_to_prop]
+  simp [hy]
+
+theorem msg_neg_neg_mod_neg {x y : BitVec w} (hx : x.msb = true) (hy : y.msb = true) :
+    (-(-x % -y)).msb = (-x % -y != 0#w && -x % -y != intMin w ^^ decide (x = intMin w) &&
+    decide (-x < -y)) := by
+  by_cases hw : w = 0; subst hw; decide +revert
+  have wpos : 0 < w := by omega
+  simp only [msb_neg]
+  simp only [msb_neg_umod_neg_of_msb_true_of_msb_true hx hy]
+
+theorem toInt_dvd_of_umod_zero {x y : BitVec w} :
+    (if x.msb then -x else x) % (if y.msb then -y else y) = 0#w ↔ y.toInt ∣ x.toInt := by
+  constructor
+  <;> by_cases hxmsb : x.msb <;> by_cases hymsb: y.msb
+  <;> intros h
+  <;> simp only [hxmsb, hymsb, reduceIte, false_eq_true, toNat_eq, toNat_umod, toNat_ofNat,
+        zero_mod, toInt_eq_neg_toNat_neg_of_msb_true, Int.dvd_neg, Int.neg_dvd,
+        toInt_eq_toNat_of_msb] at h
+  <;> simp only [hxmsb, hymsb, toInt_eq_neg_toNat_neg_of_msb_true, toInt_eq_toNat_of_msb,
+        Int.dvd_neg, Int.neg_dvd, toNat_eq, toNat_umod, reduceIte, toNat_ofNat, zero_mod]
+  <;> norm_cast
+  <;> norm_cast at h
+  <;> simp only [dvd_of_mod_eq_zero, h, dvd_iff_mod_eq_zero.mp, reduceIte]
+
+theorem toInt_smod {x y : BitVec w} :
+    (x.smod y).toInt = x.toInt.fmod y.toInt := by
+
+  rcases w with _|w ; simp [of_length_zero]
+  by_cases hxzero : x = 0#(w + 1) ; simp [hxzero]
+  by_cases hyzero : y = 0#(w + 1) ; simp [hyzero]
+
+  rw [smod_eq]
+
+  cases hxmsb : x.msb <;> cases hymsb : y.msb
+
+  <;> simp only [umod_eq]
+
+
+  · have hylt := toNat_lt_of_msb_false (x := y) (by omega)
+    have hmodlt := Nat.mod_lt (x := x.toNat) (y := y.toNat)
+      (by simp only [← toInt_inj, toInt_eq_toNat_of_msb hymsb, toInt_zero,
+        Int.natCast_eq_zero] at hyzero; omega)
+    rw [toInt_umod,
+      Int.fmod_eq_emod_of_nonneg (a := x.toInt) (b := y.toInt)
+        (by exact toInt_nonneg_of_msb_false (x := y) hymsb),
```

Comment:
```suggestion
      Int.fmod_eq_emod_of_nonneg (a := x.toInt) (b := y.toInt)
        (toInt_nonneg_of_msb_false hymsb),
```

- you can always fold away a sequence of `by exact`.
- if the arguments are clear from the assumption, you do not need to pass them explicitly

Reply:
ah I did not know, thanks!